packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+  return JSON.parse(token).token;
+}
+
+interface VercelTokenResponse {
+  token: string;
+}
+
+async function getVercelOidcToken(
+  authToken: string,
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const url = `https://api.vercel.com/projects/token?projectId=${projectId}&source=vercel-oidc-refresh`;
+    const res = await fetch(url, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${authToken}`,
+      },
+    });
+    if (!res.ok) {
+      throw new Error(`failed to refresh token: ${res.statusText}`);
+    }
+    if (!res.json) {
+      throw new Error('failed to refresh token: no json response');
+    }
+    const tokenRes = (await res.json()) as VercelTokenResponse;
+    return tokenRes;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to refresh token: ${message}`);
+  }
+}
+
+async function findProjectId(): Promise<string> {
+  const dir = await findRootDir();
+  if (!dir) {
+    throw new Error('unable to find root directory');
+  }
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    const pkgPath = path.join(dir, '.vercel', 'project.json');
+    if (!fs.existsSync(pkgPath)) {
+      throw new Error('project.json not found');
+    }
+    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
+    return pkg.projectId;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`unable to find project id: ${message}`);
+  }
+}
+
+async function findRootDir(): Promise<string> {
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    let dir = process.cwd();
+    while (dir !== path.dirname(dir)) {
+      const pkgPath = path.join(dir, '.vercel');
+      if (fs.existsSync(pkgPath)) {
+        return dir;
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new Error('token refresh only supported in node server environments');
+  }
+  throw new Error('unable to find root directory');
+}
+
+async function saveToken(
+  token: VercelTokenResponse,
+  projectId: string
+): Promise<void> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
+    );
+    const tokenJson = JSON.stringify(token);
+    const fs = await import('fs');
+    fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
+    fs.writeFileSync(tokenPath, tokenJson);
+    return;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to save token: ${message}`);
+  }
+}
+
+async function loadToken(
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const fs = await import('fs');
```

Comment:
If we could somehow replace `await import` with synchronous `require` or module-level `import`, then we could actually use this inside of `getVercelOidcTokenSync`, and try

- the request header,
- the environment variable, or
- the local disk.

Reply:
The dynamic imports will also guard for edge functions unfortunately